ts/Data/DataTable.ts

```diff
@@ -0,0 +1,902 @@
+
+        for (let i = 0, iEnd = rows.length; i < iEnd; ++i) {
+            row = rows[i];
+            row.converter = converter;
+            rowsIdMap[row.id] = row;
+            this.watchRow(row);
+        }
+    }
+
+    /* *
+     *
+     *  Properties
+     *
+     * */
+
+    /**
+     * Converter for value conversions in table rows.
+     */
+    public readonly converter: DataConverter;
+
+    /**
+     * ID of the table. As an inner table the ID links to the column ID in the
+     * outer table.
+     */
+    public id: string;
+
+    /**
+     * A map of aliases for column names
+     * [Alias]: columnName
+     */
+    public readonly aliasMap: Record<string, string>;
+
+    /**
+     * Array of all rows in the table.
+     */
+    private rows: Array<DataTableRow>;
+
+    /**
+     * Registry as record object with row IDs and their DataTableRow instance.
+     */
+    private rowsIdMap: Record<string, DataTableRow>;
+
+    /**
+     * Internal version tag that changes with each modification of the table or
+     * a related row.
+     */
+    private versionTag?: string;
+
+    /**
+     * Registry of record objects with row ID and an array of unregister
+     * functions of DataTableRow-related callbacks. These callbacks are used to
+     * keep the version tag changing in case of row modifications.
+     */
+    private watchsIdMap: Record<string, Array<Function>>;
+
+    /* *
+     *
+     *  Functions
+     *
+     * */
+
+    /**
+     * Removes all rows from this table.
+     *
+     * @param {DataEventEmitter.EventDetail} [eventDetail]
+     * Custom information for pending events.
+     *
+     * @emits DataTable#clearTable
+     * @emits DataTable#afterClearTable
+     */
+    public clear(eventDetail?: DataEventEmitter.EventDetail): void {
+
+        this.emit({ type: 'clearTable', detail: eventDetail });
+
+        const rowIds = this.getAllRowIds();
+
+        for (let i = 0, iEnd = rowIds.length; i < iEnd; ++i) {
+            this.unwatchRow(rowIds[i], true);
+        }
+
+        this.rows.length = 0;
+        this.rowsIdMap = {};
+        this.watchsIdMap = {};
+
+        this.emit({ type: 'afterClearTable', detail: eventDetail });
+    }
+
+    /**
+     * Deletes a row in this table.
+     *
+     * @param {string} rowId
+     * Name of the row to delete.
+     *
+     * @param {DataEventEmitter.EventDetail} [eventDetail]
+     * Custom information for pending events.
+     *
+     * @return {boolean}
+     * Returns true, if the delete was successful, otherwise false.
+     *
+     * @emits DataTable#deleteRow
+     * @emits DataTable#afterDeleteRow
+     */
+    public deleteRow(
+        rowId: string,
+        eventDetail?: DataEventEmitter.EventDetail
+    ): (DataTableRow|undefined) {
+        const row = this.rowsIdMap[rowId],
+            index = this.rows.indexOf(row);
+
+        this.emit({ type: 'deleteRow', detail: eventDetail, index, row });
+
+        this.rows[index] = row;
```

Comment:
Why row is still a rows array element after delete? Is it correct?

Reply:
Good question. I guess it is incorrect.

Reply:
Fixed,